internal/user_access.go

```diff
@@ -10,59 +10,52 @@ package internal
 import (
 	"errors"
 	"project-ricotta/bechamel-api/model"
-
-	"golang.org/x/exp/slices"
 )
 
 var lasagnaLoveUsers = []model.LasagnaLoveUser{
 	{ID: 1, Username: "TestUser1", Password: "password1", GivenName: "Test", FamilyName: "UserOne"},
 	{ID: 2, Username: "TestUser2", Password: "password2", GivenName: "Test", FamilyName: "UserTwo"},
 }
 
+func findUser(userFilter func(model.LasagnaLoveUser) bool) (model.LasagnaLoveUser, error) {
+	for _, user := range lasagnaLoveUsers {
+		if userFilter(user) {
+			return user, nil
+		}
+	}
+	return model.LasagnaLoveUser{}, errors.New("no user found with the supplied criteria")
+}
+
 func AuthorizeUser(userName string, password string) (model.LasagnaLoveUser, error) {
 	if userName == "" || password == "" {
 		return model.LasagnaLoveUser{}, errors.New("userName and password must be non-empty")
 	}
 
-	idx := slices.IndexFunc(lasagnaLoveUsers,
-		func(l model.LasagnaLoveUser) bool { return l.Username == userName && l.Password == password })
-	if idx == -1 {
-		return model.LasagnaLoveUser{}, errors.New("no user with supplied userName and password found")
-	}
-	return lasagnaLoveUsers[idx], nil
+	return findUser(func(u model.LasagnaLoveUser) bool {
+		return u.Username == userName && u.Password == password
+	})
 }
 
 func GetUserByID(userID int) (model.LasagnaLoveUser, error) {
-	idx := slices.IndexFunc(lasagnaLoveUsers,
-		func(l model.LasagnaLoveUser) bool { return l.ID == userID })
-	if idx == -1 {
-		return model.LasagnaLoveUser{}, errors.New("no user with supplied userID found")
-	}
-	return lasagnaLoveUsers[idx], nil
+	return findUser(func(u model.LasagnaLoveUser) bool { return u.ID == userID })
 }
 
 func GetUserByUserName(userName string) (model.LasagnaLoveUser, error) {
 	if userName == "" {
 		return model.LasagnaLoveUser{}, errors.New("userName must be non-empty")
 	}
-	idx := slices.IndexFunc(lasagnaLoveUsers,
-		func(l model.LasagnaLoveUser) bool { return l.Username == userName })
-	if idx == -1 {
-		return model.LasagnaLoveUser{}, errors.New("no user with supplied userName found")
-	}
-	return lasagnaLoveUsers[idx], nil
+
+	return findUser(func(u model.LasagnaLoveUser) bool { return u.Username == userName })
 }
 
 func AddNewUser(newUserProfile model.LasagnaLoveUser) (model.LasagnaLoveUser, error) {
-	// Not allowed to specify an userID - error if one is provided
-	if newUserProfile.ID != 0 {
-		return model.LasagnaLoveUser{}, errors.New("userID may not be specified")
+	if newUserProfile.ID != 0 || newUserProfile.FamilyName == "" || newUserProfile.GivenName == "" ||
```

Comment:
This update changes the code from not allowing an ID to be specified, to requiring an ID to be specified. Allowing callers to tell the API what the ID of added records is, is not recommended practice and can introduce security issues.
Can the previous behavior and error for when an ID is provided please be restored?
Also, why the change in error text for an incomplete record? Is this intentional?

Reply:
fixing and updating: insecure direct object references